### PR TITLE
Byzantium Fork Rules

### PR DESCRIPTION
**Pull request by pipermerriam**
### What was wrong?

Need support for the Byzantium fork.

### How was it fixed?

- [x] [EIP150 REVERT opcode](https://github.com/ethereum/EIPs/pull/206)
- [x] [EIP658 Status field in txn receipts](https://github.com/ethereum/EIPs/pull/658)
- [x] [EIP196 alt_bn128](https://github.com/ethereum/EIPs/pull/213)
- [x] [EIP197 pairing checks](https://github.com/ethereum/EIPs/pull/212)
- [x] [EIP198 big integer modular exponentiation](https://github.com/ethereum/EIPs/pull/198)
- [x] [EIP211 variable length return data](https://github.com/ethereum/EIPs/pull/211)
- [x] [EIP214 STATICCALL opcode](https://github.com/ethereum/EIPs/pull/214)
- [x] [EIP100 difficulty formula adjustment](https://github.com/ethereum/EIPs/issues/100)
- [x] [EIP649 ice age delay](https://github.com/ethereum/EIPs/pull/669)

#### Cute Animal Picture

> put a cute animal picture here.

### CODE REVIEW

File: evm/chains/chain.py [hunk 1]
```diff
@@ -122,10 +122,14 @@ def get_vm_class_for_block_number(self, block_number):
         validate_block_number(block_number)
         for n in reversed(self.vms_by_range.keys()):
             if block_number >= n:
-                return self.vms_by_range[n]
+                base_vm_class = self.vms_by_range[n]
+                break
         else:
             raise VMNotFound("No vm available for block #{0}".format(block_number))
 
+        vm_class = base_vm_class.configure(chaindb=self.chaindb)
```
Comment (pipermerriam) [inline evm/chains/chain.py]: TODO: undo this.  No longer necessary.

**Comment by pipermerriam** [comment]
*maybe* green?  This is going to get split into about 5 pull requests with some minor refactors extracted into their own pull requests.

File: evm/chains/chain.py [hunk 2]
```diff
@@ -257,6 +257,11 @@ def import_block(self, block, perform_validation=True):
 
         self.chaindb.persist_block_to_db(imported_block)
         self.header = self.create_header_from_parent(self.get_canonical_head())
+        self.logger.debug(
+            'IMPORTED_BLOCK: number %s | hash %s',
+            imported_block.number,
+            encode_hex(imported_block.hash),
+        )
```
Comment (pipermerriam) [inline evm/chains/chain.py]: TODO: extract

File: evm/constants.py [hunk 3]
```diff
@@ -46,11 +46,11 @@
...
 REFUND_SCLEAR = 15000
-REFUND_SUICIDE = 24000
+REFUND_SELFDESTRUCT = 24000
 
-GAS_SUICIDE = 0
-GAS_SUICIDE_EIP150 = 5000
-GAS_SUICIDE_NEWACCOUNT = 25000
+GAS_SELFDESTRUCT = 0
+GAS_SELFDESTRUCT_EIP150 = 5000
+GAS_SELFDESTRUCT_NEWACCOUNT = 25000
```
Comment (pipermerriam) [inline evm/constants.py]: TODO: extract

File: evm/constants.py [hunk 4]
```diff
@@ -110,7 +116,8 @@
 DIFFICULTY_MINIMUM = 131072
 
 FRONTIER_DIFFICULTY_ADJUSTMENT_CUTOFF = 13
-HOMESTEAD_DIFF_ADJUSTMENT_CUTOFF = 10
+HOMESTEAD_DIFFICULTY_ADJUSTMENT_CUTOFF = 10
```
Comment (pipermerriam) [inline evm/constants.py]: TODO: extract

File: evm/logic/block.py [hunk 5]
```diff
@@ -3,9 +3,7 @@
 
 def blockhash(computation):
     block_number = computation.stack.pop(type_hint=constants.UINT256)
-
```
Comment (pipermerriam) [inline evm/logic/block.py]: TODO: undo

File: evm/logic/call.py [hunk 6]
```diff
@@ -61,64 +65,65 @@ def __call__(self, computation):
...
         stack_too_deep = computation.msg.depth + 1 > constants.STACK_DEPTH_LIMIT
 
+        # Child message construction
+        with computation.vm.state_db(read_only=True) as state_db:
+            if code_address:
+                code = state_db.get_code(code_address)
+            else:
+                code = state_db.get_code(to)
+
+        child_msg_kwargs = {
```
Comment (pipermerriam) [inline evm/logic/call.py]: TODO: move back into `else` block

File: evm/logic/context.py [hunk 7]
```diff
@@ -64,7 +67,7 @@ def calldatacopy(computation):
     word_count = ceil32(size) // 32
     copy_gas_cost = word_count * constants.GAS_COPY
 
-    computation.gas_meter.consume_gas(copy_gas_cost, reason="Data copy fee")
+    computation.gas_meter.consume_gas(copy_gas_cost, reason="CALLDATACOPY fee")
```
Comment (pipermerriam) [inline evm/logic/context.py]: TODO: extract

File: evm/logic/system.py [hunk 8]
```diff
@@ -28,23 +30,34 @@ def return_op(computation):
...
+    start_position, size = computation.stack.pop(num_items=2, type_hint=constants.UINT256)
+
+    computation.extend_memory(start_position, size)
+
+    output = computation.memory.read(start_position, size)
+    computation.output = bytes(output)
+    raise Revert(bytes(output))
+
+
+def selfdestruct(computation):
```
Comment (pipermerriam) [inline evm/logic/system.py]: TODO: extract

File: evm/logic/system.py [hunk 9]
```diff
@@ -141,16 +154,22 @@ def __call__(self, computation):
             create_address=contract_address,
         )
 
-        child_computation = computation.vm.apply_create_message(child_msg)
-        computation.children.append(child_computation)
+        child_computation = computation.apply_child_computation(child_msg)
```
Comment (pipermerriam) [inline evm/logic/system.py]: TODO: extract

File: evm/mnemonics.py [hunk 10]
```diff
@@ -160,4 +163,5 @@
 CALLCODE = 'CALLCODE'
 RETURN = 'RETURN'
 DELEGATECALL = 'DELEGATECALL'
-SUICIDE = 'SUICIDE'
+STATICCALL = 'STATICCALL'
+SELFDESTRUCT = 'SELFDESTRUCT'
```
Comment (pipermerriam) [inline evm/mnemonics.py]: TODO: extract

File: evm/opcode_values.py [hunk 11]
```diff
@@ -181,4 +183,6 @@
 CALLCODE = 0xf2
 RETURN = 0xf3
 DELEGATECALL = 0xf4
-SUICIDE = 0xff
+STATICCALL = 0xfa
+SELFDESTRUCT = 0xff
```
Comment (pipermerriam) [inline evm/opcode_values.py]: TODO: extract

File: evm/precompile.py [hunk 12]
```diff
@@ -1,107 +0,0 @@
-import hashlib
```
Comment (pipermerriam) [inline evm/precompile.py]: TODO: extract

File: evm/precompiles/__init__.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+from .sha256 import precompile_sha256  # noqa: F401
```
Comment (pipermerriam) [inline evm/precompiles/__init__.py]: TODO: extract

File: evm/rlp/headers.py [hunk 14]
```diff
@@ -85,6 +85,12 @@ def __init__(self,
             nonce=nonce,
         )
 
+    def __repr__(self):
```
Comment (pipermerriam) [inline evm/rlp/headers.py]: TODO: extract

File: evm/rlp/receipts.py [hunk 15]
```diff
@@ -22,7 +22,7 @@
 class Receipt(rlp.Serializable):
 
     fields = [
-        ('state_root', trie_root),
+        ('state_root', binary),
```
Comment (pipermerriam) [inline evm/rlp/receipts.py]: TODO: extract

File: evm/utils/bn128.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
+from py_ecc import (
+    optimized_bn128 as bn128,
+)
+
+from evm.exceptions import (
+    ValidationError,
+)
+
+
+def validate_point(x, y):
```
Comment (pipermerriam) [inline evm/utils/bn128.py]: TODO: move to validators (maybe)

File: evm/utils/padding.py [hunk 17]
```diff
@@ -1,6 +1,9 @@
-import functools
+from cytoolz import (
+    curry,
+)
 
 
+@curry
```
Comment (pipermerriam) [inline evm/utils/padding.py]: TODO: extract

File: evm/vm/base.py [hunk 18]
```diff
@@ -7,21 +7,14 @@
 
 from evm.constants import (
     BLOCK_REWARD,
-    NEPHEW_REWARD,
```
Comment (pipermerriam) [inline evm/vm/base.py]: TODO: extract

File: evm/vm/base.py [hunk 19]
```diff
@@ -35,18 +28,15 @@ class VM(object):
...
 
     def __init__(self, header, chaindb):
-        if chaindb is None:
-            raise ValueError("VM classes must have a `db`")
-
         self.chaindb = chaindb
-
         block_class = self.get_block_class()
-        self.block = block_class.from_header(header=header, chaindb=chaindb)
+        self.block = block_class.from_header(header=header, chaindb=self.chaindb)
```
Comment (pipermerriam) [inline evm/vm/base.py]: TODO: remove

File: evm/vm/base.py [hunk 20]
```diff
@@ -138,12 +127,13 @@ def apply_computation(self, message):
                 return computation
 
             for opcode in computation.code:
-                opcode_fn = self.get_opcode_fn(opcode)
```
Comment (pipermerriam) [inline evm/vm/base.py]: TODO: extract

File: evm/vm/base.py [hunk 21]
```diff
@@ -159,7 +149,12 @@ def get_block_reward(self, block_number):
         return BLOCK_REWARD
 
     def get_nephew_reward(self, block_number):
-        return NEPHEW_REWARD
+        return self.get_block_reward(block_number) // 32
+
+    def get_uncle_reward(self, block_number, uncle):
```
Comment (pipermerriam) [inline evm/vm/base.py]: TODO: extract

File: evm/vm/computation.py [hunk 22]
```diff
@@ -133,14 +137,13 @@ def extend_memory(self, start_position, size):
...
 
-        if self.logger is not None:
-            self.logger.debug(
-                "MEMORY: size (%s -> %s) | cost (%s -> %s)",
-                before_size,
-                after_size,
-                before_cost,
-                after_cost,
-            )
+        self.logger.debug(
```
Comment (pipermerriam) [inline evm/vm/computation.py]: TODO: extract

File: evm/vm/computation.py [hunk 23]
```diff
@@ -158,21 +161,55 @@ def extend_memory(self, start_position, size):
...
     def output(self):
-        if self.error:
+        if self.error and self.error.zeros_return_data:
             return b''
         else:
             return self._output
 
     @output.setter
     def output(self, value):
+        validate_is_bytes(value)
```
Comment (pipermerriam) [inline evm/vm/computation.py]: TODO: extract

File: evm/vm/computation.py [hunk 24]
```diff
@@ -158,21 +161,55 @@ def extend_memory(self, start_position, size):
...
         self._output = value
 
+    #
+    # Runtime operations
+    #
+    def apply_child_computation(self, child_msg):
```
Comment (pipermerriam) [inline evm/vm/computation.py]: TODO: extract

File: evm/vm/computation.py [hunk 25]
```diff
@@ -158,21 +161,55 @@ def extend_memory(self, start_position, size):
...
+
+    def get_opcode_fn(self, opcode):
+        try:
+            return self.vm.opcodes[opcode]
+        except KeyError:
+            return InvalidOpcode(opcode)
+
     def register_account_for_deletion(self, beneficiary):
-        validate_canonical_address(beneficiary, title="Suicide beneficiary address")
+        validate_canonical_address(beneficiary, title="Self destruct beneficiary address")
```
Comment (pipermerriam) [inline evm/vm/computation.py]: TODO: extract

File: evm/vm/computation.py [hunk 26]
```diff
@@ -234,52 +271,59 @@ def get_gas_remaining(self):
...
-                        "gas-used: %s | gas-remaining: %s"
-                    ),
-                    encode_hex(self.msg.sender),
-                    encode_hex(self.msg.to),
-                    self.msg.value,
-                    self.msg.depth,
-                    self.msg.gas - self.gas_meter.gas_remaining,
-                    self.gas_meter.gas_remaining,
-                )
+            self.logger.debug(
```
Comment (pipermerriam) [inline evm/vm/computation.py]: TODO: extract

File: evm/vm/forks/eip150/opcodes.py [hunk 27]
```diff
@@ -39,10 +39,10 @@
         mnemonic=mnemonics.SLOAD,
         gas_cost=constants.GAS_SLOAD_EIP150,
     ),
-    opcode_values.SUICIDE: as_opcode(
-        logic_fn=system.suicide_eip150,
-        mnemonic=mnemonics.SUICIDE,
-        gas_cost=constants.GAS_SUICIDE_EIP150,
+    opcode_values.SELFDESTRUCT: as_opcode(
```
Comment (pipermerriam) [inline evm/vm/forks/eip150/opcodes.py]: TODO: extract

File: evm/vm/forks/frontier/__init__.py [hunk 28]
```diff
@@ -40,6 +39,14 @@
 )
 
 
+FRONTIER_PRECOMPILES = {
```
Comment (pipermerriam) [inline evm/vm/forks/frontier/__init__.py]: TODO: extract

File: evm/vm/forks/frontier/__init__.py [hunk 29]
```diff
@@ -50,12 +57,10 @@ def _execute_frontier_transaction(vm, transaction):
...
     vm.validate_transaction(transaction)
 
-    gas_cost = transaction.gas * transaction.gas_price
+    gas_fee = transaction.gas * transaction.gas_price
     with vm.state_db() as state_db:
-        sender_balance = state_db.get_balance(transaction.sender)
-
         # Buy Gas
-        state_db.set_balance(transaction.sender, sender_balance - gas_cost)
+        state_db.delta_balance(transaction.sender, -1 * gas_fee)
```
Comment (pipermerriam) [inline evm/vm/forks/frontier/__init__.py]: TODO: extract

File: evm/vm/forks/frontier/__init__.py [hunk 30]
```diff
@@ -251,22 +230,21 @@ def _apply_frontier_create_message(vm, message):
         contract_code = computation.output
 
         if contract_code:
-            contract_code_gas_cost = len(contract_code) * constants.GAS_CODEDEPOSIT
+            contract_code_gas_fee = len(contract_code) * constants.GAS_CODEDEPOSIT
```
Comment (pipermerriam) [inline evm/vm/forks/frontier/__init__.py]: TODO: extract

File: evm/vm/forks/frontier/blocks.py [hunk 31]
```diff
@@ -210,6 +210,31 @@ def get_cumulative_gas_used(self):
     def receipts(self):
         return self.chaindb.get_receipts(self.header, Receipt)
 
+    def make_receipt(self, transaction, computation):
```
Comment (pipermerriam) [inline evm/vm/forks/frontier/blocks.py]: TODO: extract

File: evm/vm/forks/spurious_dragon/opcodes.py [hunk 32]
```diff
@@ -23,10 +23,10 @@
         mnemonic=mnemonics.EXP,
         gas_cost=constants.GAS_EXP_EIP160,
     ),
-    opcode_values.SUICIDE: as_opcode(
-        logic_fn=system.suicide_eip161,
-        mnemonic=mnemonics.SUICIDE,
-        gas_cost=constants.GAS_SUICIDE_EIP150,
+    opcode_values.SELFDESTRUCT: as_opcode(
```
Comment (pipermerriam) [inline evm/vm/forks/spurious_dragon/opcodes.py]: TODO: extract

File: tests/core/vm/test_computation.py [hunk 33]
```diff
@@ -167,7 +161,7 @@ def test_add_log_entry_raises_if_data_isnt_in_bytes(computation):
 def test_add_log_entry(computation):
     # Adds log entry to log entries
     computation.add_log_entry(CANONICAL_ADDRESS_A, [1, 2, 3], b'')
-    assert computation.log_entries == [(b'\x0fW.R\x95\xc5\x7f\x15\x88o\x9b&>/m-l{^\xc6',
+    assert computation.log_entries == [(b'\x0fW.R\x95\xc5\x7f\x15\x88o\x9b&>/m-l\x7b^\xc6',
```
Comment (pipermerriam) [inline tests/core/vm/test_computation.py]: TODO: extract

File: tests/core/vm/test_computation.py [hunk 34]
```diff
@@ -176,43 +170,121 @@ def test_add_log_entry(computation):
...
-    computation = create_error(computation)
+    computation.add_log_entry(CANONICAL_ADDRESS_A, [1, 2, 3], b'')
+    with computation:
+        raise Revert('Triggered VMError for tests')
+    assert computation.error
     assert computation.get_log_entries() == ()
 
 
-def test_get_gas_refund(computation, create_error):
+def test_get_gas_refund(computation):
```
Comment (pipermerriam) [inline tests/core/vm/test_computation.py]: TODO:  extract

File: tests/json-fixtures/test_blockchain.py [hunk 35]
```diff
@@ -88,51 +86,58 @@ def fixture(fixture_data):
...
+    network = fixture['network']
 
-    if 'TransitionTests/bcFrontierToHomestead' in fixture_path:
+    if network == 'Frontier':
         return (
             (0, FrontierVM),
-            (5, HomesteadVM),
         )
-    elif 'TransitionTests/bcHomesteadToEIP150' in fixture_path:
+    elif network == 'Homestead':
```
Comment (pipermerriam) [inline tests/json-fixtures/test_blockchain.py]: TODO: extract

File: tests/json-fixtures/test_state.py [hunk 36]
```diff
@@ -185,8 +190,8 @@ def test_state_fixtures(fixture, fixture_vm_class):
 
     try:
         computation = vm.apply_transaction(transaction)
-    except ValidationError:
-        transaction_error = True
+    except ValidationError as err:
```
Comment (pipermerriam) [inline tests/json-fixtures/test_state.py]: TODO: extract

File: evm/logic/system.py [hunk 37]
```diff
@@ -28,6 +30,16 @@ def return_op(computation):
...
 
 
+def revert(computation):
+    start_position, size = computation.stack.pop(num_items=2, type_hint=constants.UINT256)
+
+    computation.extend_memory(start_position, size)
+
+    output = computation.memory.read(start_position, size)
+    computation.output = bytes(output)
+    raise Revert(bytes(output))
```
Comment (gsalgado) [inline evm/logic/system.py]: Maybe move the `bytes()` call to line 38 so that you don't have to repeat it in 39 and 40?

File: evm/logic/system.py [hunk 38]
```diff
@@ -147,10 +159,17 @@ def __call__(self, computation):
         if child_computation.error:
             computation.stack.push(0)
         else:
-            computation.gas_meter.return_gas(child_computation.gas_meter.gas_remaining)
             computation.stack.push(contract_address)
+        computation.gas_meter.return_gas(child_computation.gas_meter.gas_remaining)
```
Comment (gsalgado) [inline evm/logic/system.py]: Why was this change necessary?

File: evm/precompiles/ecadd.py [hunk 39]
```diff
@@ -0,0 +1,54 @@
...
+    computation.gas_meter.consume_gas(constants.GAS_ECADD, reason='ECADD Precompile')
+
+    try:
+        result = _ecadd(computation.msg.data)
+    except ValidationError:
+        raise VMError("Invalid ECADD parameters")
+
+    result_bytes = b''.join((
+        pad32(int_to_big_endian(result[0].n)),
+        pad32(int_to_big_endian(result[1].n)),
```
Comment (gsalgado) [inline evm/precompiles/ecadd.py]: Maybe unpack `result`, to make it clear how many items we expect it to have?

File: evm/precompiles/ecpairing.py [hunk 40]
```diff
@@ -0,0 +1,107 @@
...
+)
+from evm.utils.numeric import (
+    big_endian_to_int,
+)
+from evm.utils.padding import (
+    pad32,
+)
+
+
+ZERO = (bn128.FQ2.one(), bn128.FQ2.one(), bn128.FQ2.zero())
```
Comment (gsalgado) [inline evm/precompiles/ecpairing.py]: I find this an odd name choice for a 3-tuple with seemingly non-zero items, but maybe I'm missing something?

File: evm/vm/forks/byzantium/headers.py [hunk 41]
```diff
@@ -0,0 +1,78 @@
...
+)
+from evm.vm.forks.frontier.headers import (
+    create_frontier_header_from_parent,
+)
+
+
+def compute_byzantium_difficulty(parent_header, timestamp):
+    """
+    https://github.com/ethereum/EIPs/issues/100
+    TODO: figure out how to know about uncles in this context...
```
Comment (gsalgado) [inline evm/vm/forks/byzantium/headers.py]: Is this still TODO? If so, it'd be nice if you could expand on what exactly you mean

File: evm/vm/forks/byzantium/opcodes.py [hunk 42]
```diff
@@ -0,0 +1,126 @@
...
+    context,
+    logging,
+    storage,
+    system,
+)
+
+from evm.vm.forks.spurious_dragon.opcodes import SPURIOUS_DRAGON_OPCODES
+
+
+def no_static(opcode_fn):
```
Comment (gsalgado) [inline evm/vm/forks/byzantium/opcodes.py]: The name here had me confused at first, but then I realized it's used to ensure a given opcode_fn is not passed a computation with a static msg, right? Maybe rename it to `ensure_no_static()`?

File: evm/vm/forks/byzantium/opcodes.py [hunk 43]
```diff
@@ -0,0 +1,126 @@
...
+    @functools.wraps(opcode_fn)
+    def inner(computation):
+        if computation.msg.is_static:
+            raise WriteProtection("Cannot modify state while inside of a STATICCALL context")
```
Comment (gsalgado) [inline evm/vm/forks/byzantium/opcodes.py]: Also, the exception msg here is quite misleading, as it's saying we cannot modify the state when the msg is actually static? Or have I gotten this backwards?

File: evm/vm/message.py [hunk 44]
```diff
@@ -33,6 +33,7 @@ class Message(object):
     create_address = None
 
     should_transfer_value = None
+    is_static = None
```
Comment (gsalgado) [inline evm/vm/message.py]: Why does it default to None? To force all message types to define it?

File: evm/constants.py [hunk 45]
```diff
@@ -193,3 +201,11 @@
 
 # https://github.com/ethereum/EIPs/issues/170
 EIP170_CODE_SIZE_LIMIT = 24577
+
+#
+# ByzantiumVM
+#
+BYZANTIUM_MAINNET_BLOCK = 4370000
+BYZANTIUM_ROPSTEN_BLOCK = 1700000
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply evm/constants.py]: Yeah, I'm thinking it makes sense to migration the various fork-and-chain-specific block numbers into the fork or chain modules themselves.

File: evm/logic/system.py (hunk 38 above)
Comment (pipermerriam) [reply evm/logic/system.py]: The previous code with this nested into the `else` block was just an optimization since the amount of gas to return in the case of an error was always zero.  Now with the advent of revert, we should *always* return whatever gas is leftover.

File: evm/precompiles/ecpairing.py (hunk 40 above)
Comment (pipermerriam) [reply evm/precompiles/ecpairing.py]: This code is a refactored version of vitalik's implementation in pyethereum.  I'm not familiar enough with what is *actually* going on here to know what a better name would be :(

File: evm/vm/forks/byzantium/headers.py (hunk 41 above)
Comment (pipermerriam) [reply evm/vm/forks/byzantium/headers.py]: This was from before I'd put in the optimization to just check if `parent_header.uncles_hash != EMPTY_UNCLE_HASH`.

File: evm/vm/forks/byzantium/opcodes.py (hunk 43 above)
Comment (pipermerriam) [reply evm/vm/forks/byzantium/opcodes.py]: I think the txt is correct.

> "Since we are inside of a STATICCALL context, the state cannot be modified

File: evm/vm/message.py (hunk 44 above)
Comment (pipermerriam) [reply evm/vm/message.py]: Same convention that I think has raised questions previously.  This is merely a pattern I use to make it easy to know what properties an object has.  

It you have strong feelings about this pattern I'm all ears because it's something that grew organically and I'm not sure it's a good pattern.

Comment (gsalgado) [reply evm/vm/message.py]: Not really, I just find it odd for bool attributes to have a default of None, but I think it's more of a gut feeling than anything justifiable

File: evm/logic/call.py [hunk 46]
```diff
@@ -105,14 +109,18 @@ def __call__(self, computation):
...
-                computation.gas_meter.return_gas(child_computation.gas_meter.gas_remaining)
                 computation.memory.write(
                     memory_output_start_position,
                     actual_output_size,
                     child_computation.output[:actual_output_size],
                 )
-                computation.stack.push(1)
+
+            if not child_computation.error or not child_computation.error.burns_gas:
+                computation.gas_meter.return_gas(child_computation.gas_meter.gas_remaining)
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply evm/logic/call.py]: +1 but I'm going to add this as a "good first issue" (previously easy-pickings) ticket.

File: evm/logic/call.py [hunk 47]
```diff
@@ -105,14 +109,18 @@ def __call__(self, computation):
             if child_computation.error:
                 computation.stack.push(0)
             else:
+                computation.stack.push(1)
+
+            if not child_computation.error or not child_computation.error.zeros_return_data:
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply evm/logic/call.py]: +1 but going to do this as part of a "good first issue" ticket

File: evm/logic/call.py [hunk 48]
```diff
@@ -105,14 +109,18 @@ def __call__(self, computation):
             if child_computation.error:
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply evm/logic/call.py]: I don't think so or I'm not understanding your suggestion.

[the remark this answers is not in the document]
Comment (gsalgado) [reply evm/logic/call.py]: I meant `if child_computation.error is not None:`, just in case at some point we have an error class that for some reason implements `__bool__()` and returns False there, this would no longer work as expected. I can't think why we'd have an error class that would do that, but in general I prefer to explicitly check for None

[the remark this answers is not in the document]
Comment (pipermerriam) [reply evm/logic/call.py]: 👍 I'm going to file this away in a "Good First Issue" ticket.
